Refactor PostVisibility to use Popover component

aduth opened the pull request:
Related: #1193

This pull request seeks to extend generalized `Popover` behavior from #1193, refactoring `PostVisibility` to make use of the common component in place of its specific implementation. Refactoring will continue in subsequent pull requests to also revise the `PostSchedule` component. The changes here required some additional effort to improve `Popover` to be less specific to the `Inserter` use-case.

__Implementation notes:__

The `Popover` component now renders content within a nested `div`. The reason for this is to allow the root node to be 0px by 0px, positioned absolutely at the edge relative to the parent in which it is rendered. This ensures that the arrow is always shown exactly at the anchor point, and content is adjusted per the desired direction.

__Design Notes:__

Some design changes were necessary to better serve general Popover usage. Previously the left-bound inserter in post content had extended further to the left. Attempting to mirror this in the sidebar visibility dialog proved to exceed the page boundaries, so it was brought tighter in line with the node to which the dialog is anchored. Additionally, the editor header inserter Popover no longer opens anchored to the "+" icon but instead centered to the entire "+ Insert" button text.

__Testing instructions:__

While this pull request is targeted at `PostVisibility`, due to refactoring of the `Popover` component, additional testing should be done on the Inserter (small and large viewports) to ensure that no regressions have been introduced.

Repeat testing instructions from #1193.

Verify that the Post Visibility dialog appears and behaves more-or-less identical as it does in the master branch.

__Follow-up Tasks:__

- [ ] Refactor PostSchedule
- [ ] Move `onClickOutside` behavior to `Popover`, accepting an `onClose` prop handler

aduth committed:
Refactor PostVisibility to use Popover component

aduth commented on editor/sidebar/post-visibility/index.js:
```diff
@@ -106,38 +107,39 @@ class PostVisibility extends Component {
-						</div>
-						{ visibilityOptions.map( ( { value, label, info, onSelect, checked } ) => (
-							<label key={ value } className="editor-post-visibility__dialog-label">
+				<span className="editor-post-visibility__button-wrapper">
```
Majority of changes here are simply to wrap the `button` and `Popover` in a `span` to which a `position: relative;` style is applied, allowing the `Popover` to position itself within (necessary since `div` is not a valid child of `button`).

Using something like [react-portal](https://github.com/tajo/react-portal) could allow us to be less particular about the DOM structure or relative positioning. Portals are apparently supported in React 16 out-of-the-box, but I am not sure if they are documented yet.

aduth commented:
Rebased to resolve conflicts. This would be a good one to move forward on because we're having to deal with many implementation-specific requirements that would be good to consolidate (e.g. click outside, blur, etc).

jasmussen commented:
Could be due to changes recently happening in master, but I'm seeing a little weirdness with both the inserter and the visibility popover:

![screen shot 2017-07-26 at 09 01 39](https://user-images.githubusercontent.com/1204802/28608594-302a58ca-71e1-11e7-9da0-d690ff0330d7.png)

![screen shot 2017-07-26 at 09 01 58](https://user-images.githubusercontent.com/1204802/28608596-31fe656a-71e1-11e7-849f-8a45a6bca68d.png)

aduth commented:
@jasmussen I just rebased the branch, and am unable to reproduce. Could you try deleting your local copy and giving it another go? If you're still having trouble, could you let me know which browser / version you're using?

jasmussen commented:
I now see no diff. between master and this branch. Which sounds like a success criteria, so 👍 👍!

I do see this, though:

![screen shot 2017-07-28 at 09 12 32](https://user-images.githubusercontent.com/1204802/28706543-1416beae-7375-11e7-99b0-3fed55639b19.png)

But I see that also in master, and it seems to be related to #2059

aduth commented:
Currently this should only affect the visibility popover, not the date picker, but it's intended to be part of a series toward consolidating patterns to a common component (similar to DropDownMenu, #1880).

I'll plan to merge this one after today's release and address schedule popover in a subsequent pull request.

jasmussen commented:
👍👍

aduth committed:
Decrease inserter content height by tabs height

Necessary to avoid popover bounds exceed detectino from flipping inserter, moving it off-screen in mobile.

aduth commented:
In testing on mobile, I found that the inserter popover was not behaving correctly, but then I noticed the same in master. It appears that our logic for calculating the height of the inserter content was never updated when we added tabs to the inserter. This is accounted for in 5a58688.
